### DISCUSSION TITLE
How to disable all pylance rules?

### DISCUSSION THREAD

**Discussion by sobolevn** [microsoft/pylance-release#2705] [Enhancement]
Hi!

I was recently forced to switch from old MS LSP (which was amazing and served me well for like 4 years) to Pylance.
My usage pattern of such tools includes very minimum interactions of me with the tool.

I don't want any errors / warnings / etc to be shown, no noise at all. Just like it was with the MS LSP.

For now, I can disable each individual rule from https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md But, there are lots of rules inside. And probably you are going to add new ones / remove / rename existing in the future. So, it would require to sync all this things manually almost each release.

Maybe I am missing something? How can I disable it once and for all (while keeping the only feature I need: go-to-definition)? Docs have no mention of it.

**Comment by erictraut** [comment]
Most diagnostic rules in pylance are off unless you explicitly enable type checking (i.e. set `python.analysis.typeCheckingMode` to "basic"). Only the most basic of errors (those that will result in syntax errors) will appear. Those are presumably useful, since your code will crash at parse time if you try to run a source file that contains a syntax error.

The remaining diagnostic rules are all configurable, and only a few of them are enabled by default when typeCheckingMode is "off". For a full list, refer to [this table](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#diagnostic-rule-defaults). We do add new diagnostic rules occasionally, but so far all new rules have been off by default. It's unlikely that we'd add a new rule that is enabled by default, so you shouldn't need to worry about the need to continually update your settings.

It would be useful for us to understand which diagnostics you're seeing that you find annoying or unhelpful.

**Comment by sobolevn** [comment]
Thanks a lot for the quick response! 

> It would be useful for us to understand which diagnostics you're seeing that you find annoying or unhelpful.

Awesome, I would love to share my story! 🙂 

Some more context on why this is important to me.
I spend most of my day (some days even more than 12 hours) in my text editor and I really want it to be as minimalistic as possible. Here's how it looks in my ideal world:
<img width="1280" alt="Снимок экрана 2021-10-10 в 11 05 03" src="https://user-images.githubusercontent.com/4660275/136687733-f185ab2e-fab4-4486-9d1d-d1a9f87be601.png">


I've turned off everything I possibly could. All I need is code highlight and rare go-to-definition (it does not work in some cases (quite a lot of them), so I generally prefer to use `cmd+shift+f` to find symbols I need). But, still I'm used to it and sometimes right now (while MSLSP is disabled and my "languageServer" is set to "None" to keep my sanity) I miss it.

So, basically I just want to enable go-to-definition feature, that's all. But, when I enable Pylance it starts to show some output (which is might be perfectly useful for other users, just not for me). Here's how it looks like:
<img width="1280" alt="Снимок экрана 2021-10-10 в 11 04 01" src="https://user-images.githubusercontent.com/4660275/136687763-e87cd6f1-e7ad-482f-acc1-2db090e52c67.png">

What bothers me?
1. Colorful underlines
By the way, I've learned to disable them with:

```json
  "workbench.colorCustomizations": {
    // Disable wavy/squiggly underline in vscode
    "editorError.foreground":   "#00000000",
    "editorWarning.foreground": "#00000000",
    "editorInfo.foreground":    "#00000000"
  },
```

2. Red sector in right sidebar, I am still not able to disable it 😞 
3. Even when disabling all rules from https://github.com/microsoft/pyright/blob/ff55b895f98533ad9e64c635996e6eeb80129f96/packages/pyright-internal/src/common/configOptions.ts#L372-L425 I still see some errors in my message window (this is very low priority, because I almost never use it):
<img width="569" alt="Снимок экрана 2021-10-10 в 11 09 55" src="https://user-images.githubusercontent.com/4660275/136687919-cb5937c0-344e-4e42-b5a8-cdcbb6683a35.png">
4. For some reason it changes some colors of my text:
<img width="1280" alt="Снимок экрана 2021-10-10 в 11 11 39" src="https://user-images.githubusercontent.com/4660275/136687976-f2a9269d-d6a2-4fd6-83f7-db71e9ac1b92.png">

Looks like it reports unused variable / import. It still happens even after I explicitly set:

```json
"python.analysis.diagnosticSeverityOverrides": {
    "reportMissingImports": "none",
    "reportUnusedVariable": "none",
  },
```

I've also tried to disable `"editor.semanticHighlighting.enabled": false`, nothing has changed.

5. It highlights code that I write as "invalid", despite the fact I am still writing it:

Demo: https://user-images.githubusercontent.com/4660275/136688246-4bab74bd-63be-4450-86ee-e7520ac3b04c.mov

This is partially solved with hidden underlines, but side bar is still affected.

I really hope this helps to illustrate my problem. Basically, it would be a good-enough solution for me if:
1. I will be able to hide errors from my sidebar (I don't want to disable the sidebar itself, because it shows where my `git` diffs are, which is helpful)
2. I will switch off semantic highlighting of my text, like for unused variables / arguments / etc
3. Underlines hack is already working, that's fine

**Comment by jakebailey** [comment]
It sounds like the main thing you're looking for is a way to disable any and all diagnostics, something we don't have and would be pretty niche, but I believe C/C++ and Java (at least) have such a thing (and VS Code does not offer any way to disable them wholesale).

I do think there's something to be said about ensuring messages we report are under a specific diagnostic rule, though; we have a good number of diagnostics outside syntax errors that can't be disabled, which feels inconsistent.

I'll also say that having `|` or `reveal_type` is are very uncommon; `reveal_type` crashes at runtime (and is only used in the context of testing out type checking), and we only give errors on `|` if you're using an interpreter before 3.10 (when that was introduced) outside of a stub.

> It highlights code that I write as "invalid", despite the fact I am still writing it:

This is something that is explicitly a feature of all language servers I am aware of; it's what distinguishes them from the old days of on-save linters. This was largely covered by #237 (set an infinite delay so we only update things on save), in which we didn't receive wide feedback was needed.

**Comment by sobolevn** [comment]
> we don't have and would be pretty niche, but I believe C/C++ and Java (at least) have such a thing

Awesome! Maybe I can send a PR with the same feature? Would it be accepted?

> I'll also say that having | or reveal_type is are very uncommon; reveal_type crashes at runtime (and is only used in the context of testing out type checking), and we only give errors on | if you're using an interpreter before 3.10 (when that was introduced) outside of a stub.

Yes, but I am mostly working on type related things in Python, that's my main toolbox 🙂 
For example, this code is from this `typeshed` PR: https://github.com/python/typeshed/pull/6140

**Comment by erictraut** [comment]
I think we need to discuss more internally before we would consider a change like this. As Jake said, this sounds very niche. 

We have tried to consistently apply the same principles for all diagnostics in pyright/pylance. Most diagnostics are associated with diagnostic rules with configurable severity levels. A small collection of diagnostics are always emitted, but those have a close-to-zero-percent false positive rate and represent problems that will almost assuredly result in a runtime crash during parsing or during code execution. I think it's safe to say that the vast majority of developers want to be notified of these errors. You may be unique in your desire to hide this class of errors. I'm a bit mystified about why you wouldn't want to be notified of such errors in your code.

If there is a diagnostic that you think does not meet the above principle, please let us know. We would consider that a bug, and we'd be more open to fixing that.

Providing an option to disable all diagnostics is a bigger change that forces us to revisit these principles. I think we'd want to get signal from more users that this is something they'd want to see before we consider that approach.

**Comment by dmalan** [comment]
Hi @erictraut, all, just to chime in, we'd love to disable diagnostics in educational contexts, too, as red squiggles have a habit of appearing even before students have finished writing a line of code, which sometimes misleads into thinking what they've already typed is actually wrong. We're currently resorting to `"python.languageServer": "None"`, but we'd love to leave other features enabled (e.g., semantic highlighting, outlining, etc.). We tried enumerating all of https://github.com/microsoft/python-language-server/blob/master/src/Analysis/Ast/Impl/Diagnostics/ErrorCodes.cs in `python.analysis.disabled`, but errors like `Expected module name` and `Expected "import"` still slipped through, per the screenshot below. 

Are there perhaps other identifiers we overlooked that we could disable? If not, might it be compelling to support a setting akin to JavaScript's `"javascript.validate.enable": false`, a la `"python.analysis.diagnosticMode": "off"` or perhaps `"python.analysis.disabled": true`? 

Barring that, might it be possible for diagnostics to be triggered only once a user's cursor has left the line in question or the window's been blurred? At the moment, afraid the red squiggles appear so quickly as to suggest most every keystroke a student types is incorrect until their last. For first-time programmers, then, we'd ideally disable such diagnostics initially, to help them develop some muscle memory and understand errors emitted by `python` itself, and then re-enable them once they've an eye for which errors are (premature) false positives and which are true positives.

<img width="805" alt="Screen Shot 2021-10-11 at 12 04 49 PM" src="https://user-images.githubusercontent.com/788678/136821818-e8bb1de5-7ed5-4447-993e-c25c61c44dea.png">

Many thanks!

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by theRealProHacker** [comment]
I think, the option to just completely turn-off type checking while still being able to use the other cool features of the language server (go to definition (one of the best features ever invented), auto import, type info, ...) should really be a minimal requirement, IMHO. `"off"` should actually mean `"off"` and not `"off" but also not really, because we still check everything that will surely crash your program even if you just typed a single character`. That's why it would also make sense to have an option to not show (but still check for) errors, on words or lines that you're just working on, but for example still show errors that were there before you started editing and are still there. 
I might just summarize that well-implemented options are never bad.

Another good use-case for turning off linting might be when messing with `globals()`:
```python
# a small example
int_re = r"[+-]?(?:[1-9]\d*)|0" 
dec_re = fr"(?:{int_re})?(?:\.\d*)?(?:e{int_re})?"

regexes = {
    "integer": int_re,
    "number": dec_re,
    "percentage": fr"{dec_re}\%", #and so on
}
regexes: dict[str, re.Pattern] = {k:re.compile(x) for k,x in regexes.items()}
def check_regex(name: str, to_check: str)->re.Match|None:
    """
    Checks if the given regex matches the given string and returns the match
    KeyError if the regex specified doesn't exist
    """
    return regexes[name].fullmatch(to_check.strip())
for key, value in regexes.items():
    globals()[f"is_{key}"] = partial(check_regex, key)
    
if __name__ == "__main__":
    # test
    assert is_integer("3")
```
`is_integer` will be underlined with a yellow wavy line (undefined variable), even though when run it resolves to `partial(check_regex, regexes["integer"])` without a `NameError` (no close to 0% false positive). However, these yellow squiggles everywhere are **EXTREMELY** irritating. 
Right now, I think the best solution is just to write `#type: ignore` at the start of every file. Or for the above, you could just do this in `settings.json`
```json
"python.analysis.diagnosticSeverityOverrides": {
    "reportUndefinedVariable": "none"
}
```
I hope this helps someone!
